server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -0,0 +1,372 @@
+            this.missing = missing;
+        }
+
+        public void setHasScript(boolean hasScript) {
+            this.hasScript = hasScript;
+        }
+
+        public void setShowOtherBucket(boolean showOtherBucket) {
+            this.showOtherBucket = showOtherBucket;
+        }
+
+        public boolean isRewriteable(Object parent, int subAggLength) {
+            if (parent == null && subAggLength == 0 && !missing && !hasScript) {
+                if (type == Type.FILTERS) {
+                    return !showOtherBucket;
+                } else if (type == Type.DATE_HISTO) {
+                    return fieldType != null && fieldType instanceof DateFieldMapper.DateFieldType;
+                }
+            }
+            return false;
+        }
+
+        /**
+         * Different types have different pre-conditions, filter building logic, etc.
+         */
+        public enum Type {
+            FILTERS,
+            DATE_HISTO
+        }
+    }
+
+    public static long getBucketOrd(long bucketOrd) {
+        if (bucketOrd < 0) { // already seen
+            bucketOrd = -1 - bucketOrd;
+        }
+
+        return bucketOrd;
+    }
+
+    /**
+     * This should be executed for each segment
+     *
+     * @param incrementDocCount takes in the bucket key value and the bucket count
+     */
+    public static boolean tryFastFilterAggregation(
+        final LeafReaderContext ctx,
+        FastFilterContext fastFilterContext,
+        final BiConsumer<Long, Integer> incrementDocCount
+    ) throws IOException {
+        if (fastFilterContext == null) return false;
+        if (fastFilterContext.filters == null) return false;
+
+        final Weight[] filters = fastFilterContext.filters;
+        // TODO b refactor the type conversion to the context
+        final int[] counts = new int[filters.length];
+        int i;
+        for (i = 0; i < filters.length; i++) {
+            counts[i] = filters[i].count(ctx);
+            if (counts[i] == -1) {
+                // Cannot use the optimization if any of the counts
+                // is -1 indicating the segment might have deleted documents
+                return false;
+            }
+        }
+
+        int s = 0;
+        for (i = 0; i < filters.length; i++) {
+            if (counts[i] > 0) {
+                long bucketKey = i; // the index of filters is the key for filters aggregation
+                if (fastFilterContext.type == FastFilterContext.Type.DATE_HISTO) {
+                    final DateFieldMapper.DateFieldType fieldType = (DateFieldMapper.DateFieldType) fastFilterContext.fieldType;
+                    bucketKey = fieldType.convertNanosToMillis(
+                        NumericUtils.sortableBytesToLong(((PointRangeQuery) filters[i].getQuery()).getLowerPoint(), 0)
+                    );
+                }
+                incrementDocCount.accept(bucketKey, counts[i]);
+                s++;
+                if (s > fastFilterContext.size) return true;
```

Comment:
It seems the private members of static inner class are visible to the outer class. Interesting!